File: solutions/_1498_number_of_subsequences_that_satisfy_the_given_sum_condition/solution.py
from typing import List


class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        nums.sort()
        l = 0
        r = len(nums) - 1
        count = 0
        mod = 10**9 + 7

        while l <= r:
            if nums[l] + nums[r] > target:
                r -= 1
            else:
                count += pow(2, r - l, mod)
                l += 1

        return count % mod


# https://leetcode.com/problems/number-of-subsequences-that-satisfy-the-given-sum-condition/

File: solutions/_1498_number_of_subsequences_that_satisfy_the_given_sum_condition/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var numSubseq = function (nums, target) {
  nums.sort((a, b) => a - b);
  const mod = 1e9 + 7;
  let count = 0;
  let l = 0;
  let r = nums.length - 1;

  const pow = [1];
  for (let i = 1; i < nums.length; i++) {
    pow.push((pow[pow.length - 1] * 2) % mod);
  }

  while (l <= r) {
    if (nums[l] + nums[r] > target) {
      r--;
    } else {
      count = (count + pow[r - l]) % mod;
      l++;
    }
  }
  return count;
};

module.exports = { numSubseq };

/*
https://leetcode.com/problems/number-of-subsequences-that-satisfy-the-given-sum-condition/
*/
